scheduler/planner.go

```diff
@@ -195,78 +196,93 @@ type unitInfo struct {
 	} else {
-		// for mainline builds that are more recent, give them a bit
+		// For mainline builds that are more recent, give them a bit
 		// of a bump, to avoid running older builds first.
-		avgLifeTime := u.TimeInQueue / time.Duration(length)
-
-		var mainlinePriority int64
+		avgLifeTime := u.TimeInQueue / time.Duration(unitLength)
 		if avgLifeTime < time.Duration(7*24)*time.Hour {
-			mainlinePriority += u.Settings.GetMainlineTimeInQueueFactor() * int64((7*24*time.Hour - avgLifeTime).Hours())
+			breakdown.RankValueBreakdown.MainlineWaitTimeImpact = u.Settings.GetMainlineTimeInQueueFactor() * int64((7*24*time.Hour - avgLifeTime).Hours())
 		}
 		if u.ContainsStepbackTask {
-			mainlinePriority += u.Settings.GetStepbackTaskFactor()
+			breakdown.RankValueBreakdown.StepbackImpact = u.Settings.GetStepbackTaskFactor()
 		}
-
-		value += priority * mainlinePriority
 	}
-
-	// Start with the number of tasks so that units with more
-	// tasks get sorted above one-offs, and then add the priority
-	// setting as a base.
-	value += length
-	value += priority
-
-	// The remaining values are normalized per tasks, to avoid
-	// situations where larger units are always prioritized above
-	// smaller groups.
-	//
-	// Additionally, all these values are multiplied by the
-	// priority, to avoid situations where the impact of changing
-	// priority is obviated by other factors.
-
 	// Increase the value for the number of dependents, so that
 	// tasks (and units) which block other tasks run before tasks
 	// that don't block other tasks.
-	value += int64(float64(priority) * u.Settings.GetNumDependentsFactor() * float64(u.NumDependents/length))
+	breakdown.RankValueBreakdown.NumDependentsImpact = int64(u.Settings.GetNumDependentsFactor() * float64(u.NumDependents/unitLength))
 
 	// Increase the value for tasks with longer runtimes, given
 	// that most of our workloads have different runtimes, and we
 	// don't want to have longer makespans if longer running tasks
 	// have to execute after shorter running tasks.
-	value += priority * u.Settings.GetExpectedRuntimeFactor() * int64(math.Floor(u.ExpectedRuntime.Minutes()/float64(length)))
+	breakdown.RankValueBreakdown.EstimatedRuntimeImpact = u.Settings.GetExpectedRuntimeFactor() * int64(math.Floor(u.ExpectedRuntime.Minutes()/float64(unitLength)))
+
+	return 1 + breakdown.RankValueBreakdown.PatchImpact +
+		breakdown.RankValueBreakdown.PatchWaitTimeImpact +
+		breakdown.RankValueBreakdown.MainlineWaitTimeImpact +
+		breakdown.RankValueBreakdown.CommitQueueImpact +
+		breakdown.RankValueBreakdown.StepbackImpact +
+		breakdown.RankValueBreakdown.NumDependentsImpact +
+		breakdown.RankValueBreakdown.EstimatedRuntimeImpact
+}
 
-	return value
+// computePriority computes the custom priority value for this unit, which will later be multiplied with the
+// custom rank value to compute a final value by which the unit will be sorted on in the queue.
```

Comment:
Can we make a note that this destructively updates the passed in breakdown? (same with the other function)